src/hooks/useCriteriaResource.ts

```diff
@@ -0,0 +1,53 @@
+import { useCallback, useEffect, useState } from 'react';
+
+import { trpc } from '../utils/trpcClient';
+import { AddCriteriaScheme, RemoveCriteriaScheme, UpdateCriteriaScheme } from '../schema/criteria';
+
+export const useCriteriaResource = (invalidateFn: () => Promise<any>) => {
+    const [query, setQuery] = useState('');
+    const add = trpc.goal.addCriteria.useMutation();
+    const toggle = trpc.goal.updateCriteriaState.useMutation();
+    const remove = trpc.goal.removeCriteria.useMutation();
+
+    const goals = trpc.goal.suggestions.useQuery(query, {
+        enabled: query.length > 2,
+        staleTime: 0,
+        cacheTime: 0,
+    });
+
+    const onAddHandler = useCallback(
+        async (val: AddCriteriaScheme) => {
+            await add.mutateAsync(val);
+            invalidateFn();
+        },
+        [add, invalidateFn],
+    );
+    const onToggleHandler = useCallback(
+        async (val: UpdateCriteriaScheme) => {
+            await toggle.mutateAsync(val);
+            invalidateFn();
+        },
+        [invalidateFn, toggle],
+    );
+    const onRemoveHandler = useCallback(
+        async (val: RemoveCriteriaScheme) => {
+            await remove.mutateAsync(val);
+            invalidateFn();
```

Comment:
you can make a custom useCallback with invalidate and pass mutations there removing this copied code